src/path.ts

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable unicorn/prefer-at */
+export type QueryValue =
+  | string
+  | number
+  | boolean
+  | QueryValue[]
+  | Record<string, any>
+  | null
+  | undefined;
+export type QueryObject = Record<string, QueryValue | QueryValue[]>;
+
+/**
+ * Removes the leading slash from the given path if it has one.
+ */
+export function withoutLeadingSlash(path?: string): string {
+  if (!path || path === "/") {
+    return "/";
+  }
+
+  return path[0] === "/" ? path.slice(1) : path;
+}
+
+/**
+ * Removes the trailing slash from the given path if it has one.
+ */
+export function withoutTrailingSlash(path?: string): string {
+  if (!path || path === "/") {
+    return "/";
+  }
+
+  return path[path.length - 1] === "/" ? path.slice(0, -1) : path;
+}
+
+/**
+ * Joins the given base URL and path, ensuring that there is only one slash between them.
+ */
+export function joinURL(base?: string, path?: string): string {
+  if (!base || base === "/") {
+    return path || "/";
+  }
+
+  if (!path || path === "/") {
+    return base || "/";
+  }
+
+  const baseHasTrailing = base[base.length - 1] === "/";
+  const pathHasLeading = path[0] === "/";
+  if (baseHasTrailing && pathHasLeading) {
+    return base + path.slice(1);
+  }
+
+  if (!baseHasTrailing && !pathHasLeading) {
+    return `${base}/${path}`;
+  }
+
+  return base + path;
+}
+
+/**
+ * Adds the base path to the input path, if it is not already present.
+ */
+export function withBase(input = "", base = ""): string {
+  if (!base || base === "/") {
+    return input;
+  }
+
+  const _base = withoutTrailingSlash(base);
+  if (input.startsWith(_base)) {
+    return input;
+  }
+
+  return joinURL(_base, input);
+}
+
+/**
+ * Returns the URL with the given query parameters. If a query parameter is undefined, it is omitted.
+ */
+export function withQuery(input: string, query: QueryObject): string {
+  let url: URL | undefined;
+  let searchParams: URLSearchParams;
+
+  if (input.includes("?")) {
+    url = new URL(input, "http://localhost");
+    searchParams = new URLSearchParams(url.search);
+  } else {
+    searchParams = new URLSearchParams();
+  }
+
+  for (const [key, value] of Object.entries(query)) {
```

Comment:
We can also take a fast-path when there is no query in the original input. And simply use `URLSearchParams` as formatted

Reply:
I think I get it. Is it better in the new code?

Another question in this regard: Would you prefer this one to omit the empty string in favour of `undefined`?

```diff
const searchParams = new URLSearchParams(
-  searchIndex === -1 ? "" : input.slice(searchIndex + 1)
+  searchIndex === -1 ? undefined : input.slice(searchIndex + 1)
);
```

By the way, I'm using the tests from `ufo` to check against this `withQuery` implementation.